### PR TITLE
controller-manager: introduce --skipped-propagating-namespaces flag

### PR DESCRIPTION
**Pull request by pigletfly**
Signed-off-by: pigletfly <wangbing.adam@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/342

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by RainbowMango** [comment]
@pigletfly I post my comments on the referencing issue.

**Comment by pigletfly** [comment]
@RainbowMango PTAL

File: cmd/controller-manager/app/options/options.go [hunk 1]
```diff
@@ -109,4 +111,6 @@ func (o *Options) AddFlags(flags *pflag.FlagSet) {
 		"<group> for skip resources with a specific API group(e.g. networking.k8s.io),\n"+
 		"<group>/<version> for skip resources with a specific API version(e.g. networking.k8s.io/v1beta1),\n"+
 		"<group>/<version>/<kind>,<kind> for skip one or more specific resource(e.g. networking.k8s.io/v1beta1/Ingress,IngressClass) where the kinds are case-insensitive.")
+	flags.StringSliceVar(&o.SkippedPropagatingNamespaces, "skipped-propagating-namespaces", []string{},
+		"Comma-separated namespaces that should be skipped from propagating in addition to the default skipped namespaces(namespaces prefixed by kube- and karmada- and default).")
```
Comment (RainbowMango) [inline cmd/controller-manager/app/options/options.go:115]: We haven't skip the `default` namespace and have no plan on it by far.

File: pkg/controllers/namespace/namespace_sync_controller.go [hunk 2]
```diff
@@ -89,6 +91,10 @@ func (c *Controller) namespaceShouldBeSynced(namespace string) bool {
 		return false
 	}
 
+	if len(c.SkippedPropagatingNamespaces) > 0 {
+		set := sets.NewString(c.SkippedPropagatingNamespaces...)
+		return !set.Has(namespace)
+	}
```
Comment (RainbowMango) [inline pkg/controllers/namespace/namespace_sync_controller.go:97]: It's a little bit of waste to initialize a set each time.
Can we cache the namespaces by a map or set?

File: pkg/util/detector/detector.go [hunk 3]
```diff
@@ -237,6 +239,12 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
 			return false
 		}
 	}
+	// if SkippedPropagatingNamespaces is set, skip object events in the these namespaces.
+	if len(d.SkippedPropagatingNamespaces) > 0 {
+		set := sets.NewString(d.SkippedPropagatingNamespaces...)
+		return set.Has(clusterWideKey.Namespace)
+	}
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:246]: Same as above.

**Review by RainbowMango — CHANGES_REQUESTED** [review]
Some nits, otherwise LGTM.

File: cmd/controller-manager/app/options/options.go (hunk 1 above)
Comment (pigletfly) [reply cmd/controller-manager/app/options/options.go:115]: https://github.com/karmada-io/karmada/blob/master/pkg/controllers/namespace/namespace_sync_controller.go#L87

Comment (RainbowMango) [reply cmd/controller-manager/app/options/options.go:115]: It's the `namespace-sync-controller`'s default behavior, resource under `default` namespace still allowed to propagate.

File: pkg/controllers/namespace/namespace_sync_controller.go (hunk 2 above)
Comment (pigletfly) [reply pkg/controllers/namespace/namespace_sync_controller.go]: ok

**Pull request merged**